lib/client.js

```diff
@@ -1115,26 +1117,56 @@ MatrixClient.prototype.sendEvent = function(roomId, eventType, content, txnId,
         room.addPendingEvent(localEvent, txnId);
     }
 
-    if (eventType === "m.room.message" && this.sessionStore && CRYPTO_ENABLED) {
-        var e2eRoomInfo = this.sessionStore.getEndToEndRoom(roomId);
-        if (e2eRoomInfo) {
-            var encryptedContent = _encryptMessage(
-                this, roomId, e2eRoomInfo, eventType, content, txnId, callback
-            );
-            localEvent.encryptedType = "m.room.encrypted";
-            localEvent.encryptedContent = encryptedContent;
+    return _sendEvent(this, room, localEvent, callback);
+};
 
-            // TODO: Specify this in the event constructor rather than fiddling
-            // with the event object internals.
-            localEvent.encrypted = true;
-        }
+
+/**
+ * Encrypt an event according to the configuration of the room, if necessary.
+ *
+ * @param {MatrixClient} client
+ * @param {module:models/event.MatrixEvent} event  event to be sent
+ *
+ * @private
+ */
+function _encryptEventIfNeeded(client, event) {
+    if (event.isEncrypted()) {
+        // this event has already been encrypted; this happens if the
+        // encryption step succeeded, but the send step failed on the first
+        // attempt.
+        return;
     }
 
-    return _sendEvent(this, room, localEvent, callback);
-};
+    if (event.getType() !== "m.room.message") {
+        // we only encrypt m.room.message
+        return;
+    }
+
+    if (!client.sessionStore) {
+        // End to end encryption isn't enabled if we don't have a session
+        // store.
+        return;
+    }
+
+    var roomId = event.getRoomId();
+
+    var e2eRoomInfo = client.sessionStore.getEndToEndRoom(roomId);
+    if (!e2eRoomInfo || !e2eRoomInfo.algorithm) {
+        // not encrypting messages in this room
+        return;
+    }
```

Comment:
Yeah. There are some more changes afoot, which move the is-the-room-being-encrypted configuration out of the sessionstore and into the room state, which will make this look a bit different anyway. I think I'll leave this for now.